117985771 investigate collection_check_boxes usage

Maroo-b opened the pull request:
- check the use of collection_check_boxes instead of nested attributes
- useful resource: http://makandracards.com/makandra/32147-rails-4-introduced-collection_check_boxes

tansaku commented:
hi @Maroo-b - great work starting on this - I think it's a really interesting investigation

Maroo-b commented:
@tansaku I cleaned up the nested attributes management for categories by using the collection_check_box.
For the moment there is some methods in category model that are unused should we comment them for the moment or delete them ?

tansaku commented:
@Maroo-b unused methods on the model could in principle be deleted

Maroo-b commented:
Hi @tansaku, I updated the PR to remove the unused methods. The coverage dropped because I deleted a view spec used to check the category checkbox display (There is a ticket to remove all view specs https://www.pivotaltracker.com/story/show/113903069)

tansaku commented:
right, hmm - perhaps we should set up for coverage to include feature tests ... hmmm ...

tansaku commented:
great work getting this sorted @Maroo-b - I need to give it a more detailed review - sorry not to get to it today

tansaku commented:
wow @Maroo-b this is a huge step forward if this all works the way it looks like it works - it seems like we've managed to remove whole swathes of code ... I amazed that we seem to be able to lose not only our gnarly category set up code, but all the accepts nested attributes elements on the models. 

@mtc2013 @johnnymo87 @marianmosley you should definitely check out this code ... @Icicleta, you might be interested too - ooh and Tomaso - gah, can't remember his github name ...

Looking at the drop in coverage it seems like that could be fixed with the addition of an appropriate unit test for the CategoryOrganisation model:

https://coveralls.io/builds/5966538/source?filename=app%2Fmodels%2Fcategory_organisation.rb

is that something you can handle?

Maroo-b commented:
@tansaku this is more Rails approach to solve the issue. 
Sure if I have time later today I'll try to add a unit test for the PR.

Maroo-b commented on spec/models/category_organisation_spec.rb:
```diff
@@ -4,4 +4,15 @@
+      cat2 = FactoryGirl.create(:category, name: 'cat2')
+      cat_org1 = CategoryOrganisation.create(category_id: cat1.id)
+      cat_org2 = CategoryOrganisation.create(category_id: cat2.id)
+      expect_any_instance_of(Category).to receive(:<=>).with(cat2).and_call_original
```
For the unit test the `<=>` method is well covered via category model spec so to avoid duplicating the tests I added expectation that the method calls `<=>` on Category instances. 
I tried at first to use `expect(cat1).to receive(:<=>).with(cat2).and_call_original` but it didn't work 
Do you have some remarks for this test or it's acceptable ?

Maroo-b commented:
@tansaku I add a unit test for https://coveralls.io/builds/5966538/source?filename=app%2Fmodels%2Fcategory_organisation.rb 
But first is this method currently used ? because I didn't find comparison used for CategoryOrganisation records.

tansaku commented:
@Maroo-b aha - maybe that method is not used now - try removing it and see if  any tests fail? 

Also, sorry the friendly_id just got merged in so we have some possible conflicts - hopefully minimal ...

Maroo-b commented:
@tansaku I removed the unused method from CategoryOrganisation model and fixed the conflict with develop branch.

Should we add a test for https://coveralls.io/builds/6061660/source?filename=app%2Fcontrollers%2Fproposed_organisations_controller.rb#L50 ?

tansaku commented:
good job @Maroo-b - yes although really that code should be in a service rather than a private method.  I hate to make additional refactoring tasks - but this PR is a great refactoring in and of itself ... we should decide between:

1) ignoring the small coverage drop here
2) binding the feature tests in to ensure higher coverage
3) refactoring out that long private method into some domain service where the logic can be more easily tested ...

I note that there's a whole nest of stuff related to refactoring that rather smelly case statement ... hmmm ...

Maroo-b commented:
@tansaku normally this PR has resolved the related category checkbox issue, so I suggest to choose the first alternative to keep the PR concise and we can add a task for refactoring the long private method.

tansaku commented:
@Maroo-b agreed

mtc2013 commented:
Cool.
